JMH-based benchmark framework for contributors

performantdata opened the pull request:
Turns `test/benchmarks` into an SBT project that uses SBT-JMH.

This is in follow-up to [my plans mentioned on scala-internals](https://groups.google.com/forum/#!topic/scala-internals/OloyOAf3NH0).

performantdata committed:
Add JMH to the benchmark framework.
Add an example benchmark for OpenHashMap.

performantdata committed:
Add a reference to Doug Lea's benchmarks.

performantdata committed:
Add get() tests to OpenHashMap, reduce timing artifacts.

In order to get a better exploration of the variance of tests in a
limited time, I've reduced the number of measurement iterations and
increased the number of forks.  By sight, the measurement iterations
seemed pretty consistent within a trial, whereas they would vary
widely on occasional forks.

I extended testing down to 50-entry maps, to explore the rise in
service times that I was seeing at small scale.  This is probably a
timing artifact, from too-short invocations, since I'm using
@Level.Invocation in the put() tests.  To fix that, I enlarged the
unit of testing, by creating multiple, sometimes thousands, of maps
for the invocation to fill.  This has also changed the test from
filling a previously-filled map, to filling a new, but sufficiently
sized map.  The put()/remove() test now performs much worse (on a
more realistic scenario).

This also adds a couple tests for calling get() against a map that's
been filled only with put()s, or with a mix of put() and remove().

performantdata commented:
/rebuild

SethTisue commented:
presumably the integrate-ide failure is https://github.com/scala/scala-dev/issues/104 rather than anything specific to this PR

retronym commented:
Great stuff. I had toyed with this idea in https://github.com/retronym/scala-jmh-suite, which includes some plumbing to enable collection of Java Flight Recorder profiles of benchmarks.

performantdata commented:
Thanks. Thanks for the URL. I'll need to track memory usage soon. I also want to build my Gnuplot graphs automatically. More to come after this PR.

performantdata commented:
@retronym, I'm going with JOL, since it gives me easy access to memory usage from the teardown code. Can't figure how I'd do that with Flight Recorder.

retronym commented:
The more tools in the box, the better. I'm mostly after a convienient way to turn on Flight recorder when analysing benchmarks manually.

performantdata committed:
Benchmark the OpenHashMap memory usage.
Also add sbteclipse to the benchmark project.

performantdata committed:
Add a JMH runner class to the library benchmark framework.

performantdata committed:
Improve the OpenHashMapBenchmark run times.

For the warm-up invocations, suppress setup and teardown that is only
needed for the measurement iterations.  Reduce the number of forks.

performantdata commented:
/rebuild

performantdata commented:
/rebuild

performantdata commented:
/rebuild

performantdata commented:
/rebuild

adriaanm commented:
/nothingtoseehere -- The failure is due to scala/scala-dev#104, not this PR.

performantdata commented:
review by @Ichoran or @axel22 ?

performantdata commented:
@retronym would you care to review?

Ichoran commented on test/benchmarks/README.md:
```diff
@@ -0,0 +1,97 @@
+
+## running a benchmark
+
+The benchmarks require first building Scala into `../../build/pack`, using Ant.
```
Does it matter whether it's Ant or SBT?

Ichoran commented on test/benchmarks/README.md:
```diff
@@ -0,0 +1,97 @@
+`target` gets deleted on an SBT `clean`,
+so you should copy these files out of `target` if you wish to preserve them.
+
+## creating a benchmark and runner
```
Capitalize the title ("Creating a benchmark...")

Ichoran commented on test/benchmarks/README.md:
```diff
@@ -0,0 +1,97 @@
+should that be necessary for benchmarking.
+
+There are two types of classes in the source directory:
+those suffixed "`Benchmark`" and those suffixed "`Runner`".
```
Quotes and backticks are redundant.  Just use backticks.

Ichoran commented on test/benchmarks/README.md:
```diff
@@ -0,0 +1,97 @@
+This directory is a standalone SBT project, within the Scala project,
+that makes use of the [SBT plugin](https://github.com/ktoso/sbt-jmh) for [JMH](http://openjdk.java.net/projects/code-tools/jmh/).
+
+## running a benchmark
```
Capitalize first word of title (here and all following titles)

Ichoran commented on test/benchmarks/README.md:
```diff
@@ -0,0 +1,97 @@
+To show the assembly code corresponding to the code generated by the JIT compiler for specific methods,
+add `-XX:CompileCommand=print,scala.collection.mutable.OpenHashMap::*`,
+for example, to show all of the methods in the `scala.collection.mutable.OpenHashMap` class.
+If you're running OpenJDK, you may need to install the disassembler library (`hsdis-amd64.so` for the `amd64` architecture).
```
This is necessary for the Oracle JVM also (I just verified).

Ichoran commented on test/benchmarks/README.md:
```diff
@@ -0,0 +1,97 @@
+add `-XX:CompileCommand=print,scala.collection.mutable.OpenHashMap::*`,
+for example, to show all of the methods in the `scala.collection.mutable.OpenHashMap` class.
+If you're running OpenJDK, you may need to install the disassembler library (`hsdis-amd64.so` for the `amd64` architecture).
+In Debian, this is available in <a href="https://packages.debian.org/search?keywords=libhsdis0-fcml">the `libhsdis0-fcml` package</a>.
```
This will work for the OpenJDK, but you have to move the file to the right path to get it to work for the Oracle JVM.

performantdata replied:
What does "move the file to the right path" mean precisely? What should I tell them?

Ichoran replied:
"If you are using the Oracle JVM or another compatible JVM that is not included with your distribution, you may need to copy or link the `hdis-amd64.so` file from its installation location for an OpenJDK install to the `jre/lib/amd64` directory inside the Java installation that is being used."

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,111 @@
+        f.format("# [%s]\n", label)
+
+        val isMemoryUsageDataset = label.contains(memoryDatasetQualifer)
+        dataset.foreach { result ⇒
```
Use two characters for `=>`.  The occasional problems with ⇒ are not worth it.  (People not set up to type it, random unicode mangling, etc..)

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,111 @@
+          if (isMemoryUsageDataset) {
+            val memoryResult = secondaryResults.get(memoryAuxCounterName)
+            val entriesResult = secondaryResults.get(entriesAuxCounterName)
+            f.format("%s %f %f %f %f\n", size,
```
This would be a lot easier to follow if you used plain `f.println` and the f- or s-interpolator to generate the results.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,111 @@
+            val memoryResult = secondaryResults.get(memoryAuxCounterName)
+            val entriesResult = secondaryResults.get(entriesAuxCounterName)
+            f.format("%s %f %f %f %f\n", size,
+              Double.box(entriesResult.getScore), Double.box(entriesResult.getStatistics.getStandardDeviation),
```
Can you factor this pattern out?  At least `getStatistics.getStandardDeviation` could be a call to a local `std` method, but you might just make a `statStr` method or something that  does the whole `f"${x.getScore} ${x.getStatistics.getStandardDeviation}"` thing.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapBenchmark.scala:
```diff
@@ -0,0 +1,204 @@
+      map.put(i, i)
+      i += 1
+    }
+  }
```
You end here without doing anything with your constructed map.  This is generally safe enough, but it's probably good practice to include a dump of the result to Blackhole (e.g. the size) to be sure that the work isn't optimized away.  (Especially if someone copies and pastes this code for another benchmark where the putting is changed into a more elidable operation.)

performantdata replied:
The better place to blackhole results is in the `@Benchmark` methods; these are here as utility methods. But I believe that the maps being passed into those methods as external state prevents them from being optimized away.

The difficulty of eliding `put` calls wouldn't prevent the whole map from being optimized away. Conversely, if the `put` were changed to `get`, blackholing a `map.size()` would make the `get` no less elidable. I guess I don't have much sympathy either for someone who's copy-pasting the wrong example, e.g., `put_Int()` instead of `get_Int()`.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapBenchmark.scala:
```diff
@@ -0,0 +1,204 @@
+
+      base += blockSize
+    }
+  }
```
Blackholing would probably be good again.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapBenchmark.scala:
```diff
@@ -0,0 +1,204 @@
+  def get_Int(map: OpenHashMap[Int,Int], size: Int, bh: Blackhole) {
+    var i = 1
+    while (i <= size) {
+      bh.consume(map.get(i).getOrElse(0))
```
I think you'll have less overhead if you accumulate a running sum and push that into `bh` instead.

performantdata replied:
Maybe so, but I plan to test other type parameters than numerics, and would like the tests to be comparable.

retronym replied:
Or perhaps, put the `Option[Int]` into the blackhole, rather than calling `Option#getOrElse`. Or is the intention to test the combined performance of `OHM#get` / `Option#getOrElse`?

It is also worth thinking about what assumptions these benchmarks make about the scalac compiler options used to compile the tests themselves. Under `-Yopt:l:classpath`, `Option#getOrElse` is inlined 

```
% ~/scala/2.12/bin/scala -Yopt:l:classpath
Welcome to Scala 2.12.0-M4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_71).
Type in expressions for evaluation. Or try :help.

scala> class OptInline { def foo(a: Option[Int]) = a.getOrElse(0) }
defined class OptInline

scala> :javap -c OptInline#foo
  public int foo(scala.Option<java.lang.Object>);
    Code:
       0: aload_1
       1: ifnonnull     6
       4: aconst_null
       5: athrow
       6: aload_1
       7: invokevirtual #27                 // Method scala/Option.isEmpty:()Z
      10: ifeq          23
      13: invokestatic  #31                 // Method $line4$$read$$iw$$iw$OptInline$$$anonfun$1:()I
      16: invokestatic  #37                 // Method scala/runtime/BoxesRunTime.boxToInteger:(I)Ljava/lang/Integer;
      19: astore_2
      20: goto          28
      23: aload_1
      24: invokevirtual #41                 // Method scala/Option.get:()Ljava/lang/Object;
      27: astore_2
      28: aload_2
      29: invokestatic  #45                 // Method scala/runtime/BoxesRunTime.unboxToInt:(Ljava/lang/Object;)I
      32: ireturn
```

```
 ~/scala/2.12/bin/scala
Welcome to Scala 2.12.0-M4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_71).
Type in expressions for evaluation. Or try :help.

scala> class OptInline { def foo(a: Option[Int]) = a.getOrElse(0) }
defined class OptInline

scala> :javap -c OptInline#foo
  public int foo(scala.Option<java.lang.Object>);
    Code:
       0: aload_1
       1: invokedynamic #43,  0             // InvokeDynamic #0:apply$mcI$sp:()Lscala/runtime/java8/JFunction0$mcI$sp;
       6: invokevirtual #49                 // Method scala/Option.getOrElse:(Lscala/Function0;)Ljava/lang/Object;
       9: invokestatic  #55                 // Method scala/runtime/BoxesRunTime.unboxToInt:(Ljava/lang/Object;)I
      12: ireturn
```

As a starting point, I would suggest adding `-Yopt:l:classpath` to the `scalacOptions` for this project. Perhaps we'd like to parameterize this down the track, or maybe have separate sub projects with different optimization levels so we could have fine grained control over the way that benchmark code is compiled.

performantdata replied:
> is the intention to test the combined performance of OHM#get / Option#getOrElse?

Yes, because I think that's a more common usage, to subsequently unpack the `Option`.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapBenchmark.scala:
```diff
@@ -0,0 +1,204 @@
+class OpenHashMapBenchmark {
+  import OpenHashMapBenchmark._
+
+  @Param(Array("50", "100", "250", "1000", "2500", "10000", "25000", "100000", "250000", "1000000", "2500000",
```
Unless you declare that a map is only to be used for big collections, it's also valuable to test the smaller sizes.  This generally involves making an array of maps and doing something to each of them, as there is too much overhead to measure accurately the timings in a map of e.g. 2 elements.  You can just run all your benchmarks that way, since it's perfectly okay to have an array of length 1.

performantdata replied:
That's how the tests are constructed already. See the `BulkPutState` class. I'd only need to add more `size` values. But the values are _sort of_ limited to multiples of 25 by the current `blocks` value in `put_remove_Int()`.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,111 @@
+  private[this] val outputFile = new File(outputDirectory, "OpenHashMap.dat")
+
+  /** Qualifier to add to the name of a memory usage data set. */
+  private[this] val memoryDatasetQualifer = " memory"
```
Do you mean to have a space in front here?  Is it _necessary_?  If it's not necessary but you need a separator, I'd go for `-` or `_` instead, simply because it otherwise looks like an error.

Ichoran replied:
Also, "Qualifier" is misspelled.

performantdata replied:
> Do you mean to have a space in front here?

Yes. I label each dataset with its corresponding method name, which can't (wouldn't normally) contain a space. So a `-` would also work.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,111 @@
+  private[this] val memoryDatasetQualifer = " memory"
+
+  /** Name of the JMH parameter for the number of map entries per invocation. */
+  private[this] val sizeParamName = "size"
```
Is this actually accomplishing anything except bulking up the code?  Abstracting out constants is fine if you might get them wrong or they might change, but this is unlikely to change and hard to get wrong.  Also, you're not really abstracting the common operation here (which is `getParam.getParam("size")`).  Same deal for all the names: they're used to accomplish something, so abstract the action rather than just make the name longer.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,111 @@
+  
+    val opts = new CommandLineOptions(args: _*)
+    var builder = new OptionsBuilder().parent(opts)
+      .jvmArgsPrepend("-Xmx6000m")
```
There is no reason to put this on a separate line

performantdata replied:
I was previously overriding other options as well.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,111 @@
+
+    def addToDataset(result: RunResult, key: String): Unit =
+      datasetByName.get(key)
+        .getOrElse({ val d = SortedSet.empty(ordering); datasetByName.put(key, d); d }) += result
```
I generally prefer dot-on-previous-line because it works better when e.g. pasting into the REPL.  The indentation is already plenty of indication that it's a continuation of the previous command.  Also, the `+= result` bit is so far to the right it's easy to miss.  I'd just assign a temporary variable.  And are you not using `getOrElseUpdate` for some reason?  That might make it short enough so a single line is fine.

performantdata committed:
Address JMH benchmark reviewer's issues.

Besides tweaks to the documentation, this tests smaller (25-element)
maps, and rewrites OpenHashMapRunner in more idiomatic Scala.

performantdata committed:
Enable full compiler optimizations in JMH benchmarking.

performantdata commented:
I just rebased to pick up #5139.

performantdata commented:
@SethTisue, any idea what's wrong with the 2.11 build now?

adriaanm commented:
It looks like our CI was really overloaded because each commit was being validated at the same time and you pushed about 10 of them :-)

adriaanm commented:
We have limits on how many jobs Jenkins runs at the same time. Looks like they could do with some more tuning.

adriaanm commented:
Hmm, at least that was my first impression. Looks like several commits took >40min just to build the compiler. Not sure what's going on, but I haven't followed the history of this PR.

adriaanm commented:
I've added the suffix to tell the CI to only build the last commit, to help in load balancing. Let's do a rebuild when the other commits are done.

performantdata commented:
/rebuild

performantdata commented:
@adriaanm, it's still red. Do we ignore those commits, or how do I rebuild just them?

adriaanm commented:
/rebuild

adriaanm commented:
/rebuild 56c5d92

adriaanm commented:
I removed the suffix that tells the CI to only build the last commit and then `/rebuild`. Hopefully all-green in a couple hours. ⏳

performantdata commented:
Oh, it's right [there](https://github.com/scala/scabot/blob/master/README.md#commands).

performantdata commented:
Green. Back to you, @Ichoran.

performantdata commented:
@Ichoran, are you still reviewing this, Rex?

Ichoran commented:
@performantdata - I'm snowed under with commitments but I'll try to get to it sometime this weekend (probably Sunday afternoon).

performantdata commented:
@Ichoran OK thanks. It was just that I'm having to cherry-pick these commits whenever I want to test a branch, so I can just keep doing that until you're free, not a big deal.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,105 @@
+import java.io.OutputStreamWriter
+import java.io.PrintWriter
+import scala.collection.JavaConversions
+import scala.language.existentials
```
This import is kind of lost here, which is contrary to the point of requiring the import in the first place.  Can you move it to the top?

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,105 @@
+        addToDataset(result.label + memoryDatasetQualifier, result)
+    }
+
+    //TODO Write out test parameters
```
If we want to do this, can we do it?  If not, can we decide not and remove or change the comment?  (E.g. "If we wanted to write out the test parameters, we'd do here / like this: ...")

performantdata replied:
This benchmark is still a work in progress. I have a Gnuplot script that I use that ought to become part of the toolkit, but haven't decided how to incorporate it yet. Some parameters would be written out to some file for inclusion in that script. This is a reminder to me.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapRunner.scala:
```diff
@@ -0,0 +1,105 @@
+        val isMemoryUsageDataset = label.endsWith(memoryDatasetQualifier)
+        dataset.foreach { r =>
+          f.println(r.size + " " + (
+            if (isMemoryUsageDataset)
```
Can you have one line for the if and one for the else so the logic is easier to follow at a glance?  Or maybe break out the entire thing as a helper method?  `def reportStats = ...`

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapBenchmark.scala:
```diff
@@ -0,0 +1,204 @@
+
+  /** Test getting elements from a map of `Int` to `Int` from which elements have been removed. */
+  @Benchmark
+  def put_remove_get_Int(state: BulkRemovedGetState, bh: Blackhole) = OpenHashMapBenchmark.get_Int(state.map, size, bh)
```
Is this the wrong call here?  Also, can you put the result on a second line so it doesn't disappear off the right margin?

``` scala
def put_blah =
  OHMB.blah(blah)
```

performantdata replied:
> Is this the wrong call here?

No, it's correct. It's testing a `get()` call under different circumstances.

Ichoran replied:
Can you try to do something with the method names to make it clearer that that is the intention?  `get_Int_on_removed` or something like that?  You do try to explain it in the comment, but it is not terribly obvious that the `OpenHashMapBenchmark.get_Int` call is correct.

performantdata replied:
> can you put the result on a second line so it doesn't disappear off the right margin?

Admittedly it's a long line, but it fits in my IDE window. Have a specific right margin guideline, then?

Ichoran replied:
I have it be pretty ad-hoc.  My rule is, "The more important it is, the less far to the right it should be."

In this case, what you're actually benchmarking is as important as anything, so it ought not get too far out there IMO.

performantdata replied:
> Can you try to do something with the method names to make it clearer that that is the intention?

I'll address this when my outstanding PRs are in. Currently I have a number of data files hanging around, waiting on these PRs to close, that contain these method names, which are also in the Gnuplot script.

Ichoran replied:
There's no reason to make the commit history of the main repository messy just because you happen to have some open PRs now.  If those PRs are waiting on this, just branch.  If you need to revert, apply the reversion to the branch after rebasing on this.

(If you don't think the names should ever change, that's a different issue.)

performantdata replied:
I'll be adding other key types, probably other value types, to the benchmarks. This isn't the last time I'll be editing this file.

> If those PRs are waiting on this

They're not. For other branches I have these files sitting around untracked. But I've generated benchmark data files containing these method names, for those PRs, and intend to keep the data until the PRs close.

> If you don't think the names should ever change

No, I agree with that, and most of your other personal syntactic preferences. I just don't think that should hold up a PR.

Ichoran replied:
I agree with not holding up the PR.  My thought was that my attention is very limited, but I did spend some of it here, so why not take care of the problems now (if you agree there are improvements to be made)?  (Possible reason: because someone else will be reviewing the other PRs, so you won't be held back by how busy I am.)

performantdata replied:
I'm working on the follow-up PR to #5121 right now, which is more time-sensitive.

Ichoran commented on test/benchmarks/src/main/scala/scala/collection/mutable/OpenHashMapBenchmark.scala:
```diff
@@ -0,0 +1,204 @@
+class OpenHashMapBenchmark {
+  import OpenHashMapBenchmark._
+
+  @Param(Array("25", "50", "100", "250", "1000", "2500", "10000", "25000", "100000", "250000", "1000000", "2500000",
```
I'm all for not adhering to margin widths set by 1980s hardware, but since you're wrapping the list of numbers anyway, can you do it more evenly?

Ichoran commented:
Getting close!  I think there's just one real bug--or maybe it's just unclear, in which case you should leave a comment explaining why the right method is being benchmarked.  The rest is just random style stuff.  The last step will be to squash the commits into one; there doesn't seem to be any particular value to leaving them separate is there?  (You might want to leave a branch with them separate in your repository, but it seems unnecessary for the main one.)

performantdata commented:
So, it's not a bug. I can address the style stuff in a subsequent commit in the PRs that I'll soon be benchmarking but haven't opened yet. Otherwise...when would you finish review?

Ichoran commented:
Anyway, the tests look technically correct.  There doesn't seem to be any reason to keep the separate commits, so you should squash them for the PR.  The style could use a little improvement, and I'm a little wary of a promissory note instead of an actual fix (at least for most of them--the single `get_Int` call is probably fine, though again, I'm not sure the reason to not change it is very compelling).  But if someone wants to merge this on technical soundness, it's ready.

Ichoran commented:
And thank you for writing a JMH harness for testing library performance!  It's long overdue in my opinion.

performantdata commented:
LGTRex

Ichoran commented:
It almost does, but not completely, which is why I avoided the normal phrase.  It would look good if the last few issues were fixed; as it is it's technically sound so could be merged anyway.

retronym commented:
I'm going to merge this one. Thanks for this effort, @performantdata. I think the commit history is clean enough to merge without squashing in this instance.

performantdata commented:
Uh...OK, thanks. I'm currently working on the fixes that Rex asked for, and adding `AnyRef` key tests. Another PR, I guess.

Thanks, Rex!

SethTisue commented:
we don't normally release-note contributor-facing things, but perhaps we really want to draw attention to this? not sure, untag if you disagree

SethTisue commented:
>  I also want to build my Gnuplot graphs automatically. More to come after this PR.

@performantdata @szeiger well I know quite some time has passed, but I'm wondering if you, or anyone else watching the repo, has anything that does this. even if it's not fully automatic, any sort of instructions I could follow. I'd love to be able to generate nice graphs like Stefan did over at https://github.com/scala/scala/pull/8534

SethTisue commented:
ah, @retronym has directed me towards https://github.com/lightbend/benchdb
